README.md
# README

This is an HTTP Protocol Adapter for the Droonga Engine.

## Usage

### Server

If both this and fluentd are running on the same machine,
you can run this easily.

    $ npm install -g droonga-http-server
    $ droonga-http-server

Otherwise, you have to specify pairs of host and port to send messages
to the fluentd and to receive messages from the fluentd.

    $ droonga-http-server \
        --droonga-engine-host-name "backend.droonga.example.org" \
        --droonga-engine-port 24224 \
        --receive-host-name "frontend.droonga.example.org"

### Client

Frontend applications can call HTTP APIs to access resources stored in
the droonga. For example:

    GET /droonga/tables/entries?query=foobar HTTP/1.1

It works as a search request, and a JSON string will be returned as the result.

## License

The MIT License. See LICENSE for details.

Copyright (c) 2014 Droonga project


NEWS.md
# News

## 1.1.2 2015-04-29

 * Bump required version of `express-droonga`.

## 1.1.1 2014-12-02

 * Fix mismatched dependency about `express-droonga`.

## 1.1.0 2014-12-01

 * Connects to active engine nodes of the cluster of the associated engine node correctly.
   Previous version unexpectedly kept connections to already unjoined engine nodes.

## 1.0.9: 2014-11-29

 * A new `--host` option is available to restrict the listening IP address.
   The default value is `0.0.0.0` (meaning `listen all IP addresses`).
 * A new `--cache-ttl-in-seconds` option is available to set the time to live of cached responses, in seconds.
   The default value is `60` (meaning `1 minute`).
 * A new `--disable-trust-proxy` option is available to disable the feature
   even if it is activated by the static configuration file.
 * A new `--document-root` option is available to specify the document root.
   It is the path to the built-in Groonga admin page by default.
 * The `daemon` option is now ignored in the static configuration file.
   Now, you always have to specify `--daemon` option for the `droonga-http-server` command
   to start it as a daemon.
 * The `droonga-http-server-configure` command now shows prompts for all options always.
 * Responses for most commands are never cached.
   Now, only responses based on `search` or Groonga's `select` commands and the administration page are cached.
 * A new endopoint `/cache` is introduced to clear all response caches.
   To clear cached contents, send an HTTP `DELETE` request to the path.
 * Supports multiple Droonga Engine nodes as its backends.
   Now droonga-http-server can work like a load balancer.
 * The list of connecting Droonga Engine nodes is automatically updated
   based on the actual list of active members in the cluster.

## 1.0.8: 2014-10-07

 * Works correctly as a service even if you restarted the computer itself.
 * Log level of the system log is now customizable.
   You just have to put a line like `system_log_level: debug` into the `droonga-http-server.yaml`.
   And, `droonga-http-server-configure` also asks the log level.

## 1.0.7: 2014-09-29

 * The installation script is now available.
   It automatically installs required softwares and configure the `droonga-http-server` as a system service.
   Currently it works only for Debian, Ubuntu, and CentOS 7.
 * The service works as a process belonging to a user `droonga-http-server` who is specific for the service.
   The configuration directory for the service is placed under the home directory of the user.
 * A static configuration file to define default parameters (`port` and so on) is now available.
   It must be placed into the configuration directory specified by the environment variable `DROONGA_BASE_DIR`.
   You don't have to run `droonga-http-server` command with many options, anymore.
 * A new command line utility `droonga-http-server-configure` is available.
   It generates the static configuration file for the service.
 * Cached responses are now returned correctly.

## 1.0.6: 2014-07-29

 * Provides Groonga's administration page as the document root (`/`) experimentally.
 * Supports a new `--environment` option to override the `NODE_ENV` environment variable.

## 1.0.5: 2014-05-29

 * Use `Default` as the name of the default dataset.
   It is same to Droonga Engine's one.
 * Use `10041` as the default port number.
   It is same to Groonga HTTP server's one.

## 1.0.4: 2014-04-29

 * Works with the [Express 4.0](http://expressjs.com/).
 * Supports a new `--enable-trust-proxy` option to run the server behind a reverse proxy.
 * Supports a new `--plugins` option to choose plugins to be activated.
 * Supports new `--daemon` and `--pid-file` options for the daemon mode.

## 1.0.3: 2014-03-29

 * Fix broken dependencies.

## 1.0.2: 2014-03-29

 * Register bin/droonga-http-server command correctly.

## 1.0.1: 2014-03-29

The first release!


install.sh
# Copyright (C) 2014-2015 Droonga Project
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License version 2.1 as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

# Usage:
#
#  Ubuntu:
#
#   Install a release version:
#     $ curl https://raw.githubusercontent.com/droonga/droonga-http-server/master/install.sh | sudo bash
#   Install the latest revision from the repository:
#     $ curl https://raw.githubusercontent.com/droonga/droonga-http-server/master/install.sh | sudo VERSION=master bash
#   Install with specified hostnames (disabling auto-detection):
#     $ curl https://raw.githubusercontent.com/droonga/droonga-http-server/master/install.sh | sudo HOST=xxx.xxx.xxx.xxx ENGINE_HOST=xxx.xxx.xxx.xxx bash
#
#  CentOS 7:
#
#   Install a release version:
#     # curl https://raw.githubusercontent.com/droonga/droonga-http-server/master/install.sh | bash
#   Install the latest revision from the repository:
#     # curl https://raw.githubusercontent.com/droonga/droonga-http-server/master/install.sh | VERSION=master bash
#   Install with specified hostnames (disabling auto-detection):
#     # curl https://raw.githubusercontent.com/droonga/droonga-http-server/master/install.sh | HOST=xxx.xxx.xxx.xxx ENGINE_HOST=xxx.xxx.xxx.xxx bash

NAME=droonga-http-server
DOWNLOAD_URL_BASE=https://raw.githubusercontent.com/droonga/$NAME
REPOSITORY_URL=https://github.com/droonga/$NAME.git
USER=$NAME
GROUP=droonga
DROONGA_BASE_DIR=/home/$USER/droonga
TEMPDIR=/tmp/install-$NAME

EXPRESS_DROONGA_REPOSITORY_URL=git://github.com/droonga/express-droonga.git#master

: ${VERSION:=release}
: ${HOST:=Auto Detect}
: ${PORT:=10041}
: ${ENGINE_HOST:=Auto Detect}
: ${ENGINE_PORT:=Auto Detect}

case $(uname) in
  Darwin|*BSD|CYGWIN*) sed="sed -E" ;;
  *)                   sed="sed -r" ;;
esac

ensure_root() {
  if [ "$EUID" != "0" ]; then
    echo "You must run this script as the root."
    exit 1
  fi
}

guess_platform() {
  if [ -e /etc/debian_version ] || [ -e /etc/debian_release ]; then
    echo "debian"
    return 0
  elif [ -e /etc/centos-release ]; then
    echo "centos"
    return 0
  fi
  return 1
}

exist_command() {
  type "$1" > /dev/null 2>&1
}

exist_all_commands() {
  for command in $@; do
    if ! exist_command $command; then
      return 1
    fi
  done
  return 0
}

exist_yum_repository() {
  if ! yum --enablerepo=$1 repolist; then
    return 1
  fi
  yum --enablerepo=$1 repolist | grep --quiet "$1"
}

exist_user() {
  id "$1" > /dev/null 2>&1
}

prepare_user() {
  echo ""
  echo "Preparing the user..."

  groupadd $GROUP

  if ! exist_user $USER; then
    useradd -m $USER
  fi

  usermod -G $GROUP $USER
  return 0
}

detect_engine_config() {
  local config_key=$1
  local config_value=""
  local engine_config="/home/droonga-engine/droonga/droonga-engine.yaml"
  if [ -e $engine_config ]; then
    config_value=$(cat $engine_config | grep -E "^ *$config_key *:" | \
                   cut -d ":" -f 2 | $sed -e "s/^ +| +\$//g")
  fi
  echo $config_value
}

setup_configuration_directory() {
  echo ""
  echo "Setting up the configuration directory..."

  [ ! -e $DROONGA_BASE_DIR ] &&
    mkdir $DROONGA_BASE_DIR

  local config_file="$DROONGA_BASE_DIR/$NAME.yaml"
  if [ ! -e $config_file ]; then
    local should_reconfigure_engine_host="false"
    local should_reconfigure_host="false"

    if [ "$ENGINE_HOST" = "Auto Detect" ]; then
      ENGINE_HOST=$(detect_engine_config host)
      if [ "$ENGINE_HOST" != "" ]; then
        echo "The droonga-engine service is detected on this node."
        echo "The droonga-http-server is configured to be connected"
        echo "to this node ($ENGINE_HOST)."
      else
        if [ "$ENGINE_HOST" = "" ]; then
          ENGINE_HOST=$(hostname)
          should_reconfigure_engine_host="true"
        fi
        echo "This node is configured to connect to the droonga-engine node $ENGINE_HOST."
      fi
    fi

    if [ "$ENGINE_PORT" = "Auto Detect" ]; then
      ENGINE_PORT=$(detect_engine_config port)
      if [ "$ENGINE_PORT" != "" ]; then
        echo "The droonga-engine service is detected on this node."
        echo "The droonga-http-server is configured to be connected"
        echo "to this node with the port $ENGINE_PORT."
      else
        ENGINE_PORT=10031
      fi
    fi

    [ "$HOST" = "Auto Detect" ] &&
      HOST=$(determine_hostname)

    if [ "$HOST" = "" ]; then
      HOST=$(hostname)
      should_reconfigure_host="true"
    fi
    echo "This node is configured with a hostname $HOST."

    if [ "$should_reconfigure_engine_host" = "true" -o \
         "$should_reconfigure_host" = "true" ]; then
      echo "********************** CAUTION!! **********************"
      echo "Installation process coudln't detect following parameters:"
      echo ""
      if [ "$should_reconfigure_engine_host" = "true" ]; then
        echo " * the hostname of the droonga-engine node to be connected"
      fi
      if [ "$should_reconfigure_host" = "true" ]; then
        echo " * the hostname of this node, which is accessible from "
        echo "   other nodes"
      fi
      echo ""
      echo "You may have to configure droonga-http-server manually,"
      echo "by following command line:"
      echo ""
      echo "  droonga-http-server-configure --reset-config"
      echo "*******************************************************"
    fi

    # we should use --no-prompt instead of --quiet, for droonga-http-server 1.0.9 and later.
    droonga-http-server-configure --quiet \
                                  --droonga-engine-host-name=$ENGINE_HOST \
                                  --droonga-engine-port=$ENGINE_PORT \
                                  --receive-host-name=$HOST \
                                  --port=$PORT
    if [ $? -ne 0 ]; then
      echo "ERROR: Failed to configure $NAME!"
      exit 1
    fi
  fi

  chown -R $USER:$GROUP $DROONGA_BASE_DIR
}


guess_global_hostname() {
  if hostname -d > /dev/null 2>&1; then
    local domain=$(hostname -d)
    local hostname=$(hostname -s)
    if [ "$domain" != "" ]; then
      echo "$hostname.$domain"
      return 0
    fi
  fi
  echo ""
  return 1
}

determine_hostname() {
  local global_hostname=$(guess_global_hostname)
  if [ "$global_hostname" != "" ]; then
    echo "$global_hostname"
    return 0
  fi

  local address=$(hostname -i | \
                  $sed -e "s/127\.[0-9]+\.[0-9]+\.[0-9]+//g" \
                       -e "s/[0-9a-f:]+%[^ ]+//g" \
                       -e "s/  +/ /g" \
                       -e "s/^ +| +\$//g" |\
                  cut -d " " -f 1)
  if [ "$address" != "" ]; then
    echo "$address"
    return 0
  fi

  echo ""
  return 1
}


use_master_express_droonga() {
  mv package.json package.json.bak
  cat package.json.bak | \
    $sed -e "s;(express-droonga\"[^:]*:[^\"]*\")[^\"]+;\1$EXPRESS_DROONGA_REPOSITORY_URL;" \
    > package.json
}

install_from_repository() {
  cd $TEMPDIR

  if [ -d $NAME ]
  then
    cd $NAME
    git reset --hard
    git pull --rebase
    git checkout $VERSION
    use_master_express_droonga
    npm update
  else
    git clone $REPOSITORY_URL
    cd $NAME
    git checkout $VERSION
    use_master_express_droonga
  fi
  npm install -g
  rm package.json
  mv package.json.bak package.json
}

download_url() {
  if [ "$VERSION" != "release" ]; then
    echo "$DOWNLOAD_URL_BASE/master/$1"
  else
    echo "$DOWNLOAD_URL_BASE/v$(installed_version)/$1"
  fi
}

installed_version() {
  $NAME --version
}



# ====================== for Debian/Ubuntu ==========================
prepare_environment_in_debian() {
  apt-get update
  apt-get install -y curl nodejs nodejs-legacy npm

  if [ "$VERSION" != "release" ]; then
    apt-get install -y git
  fi
}
# ====================== /for Debian/Ubuntu =========================



# ========================= for CentOS 7 ============================
prepare_environment_in_centos() {
  if ! exist_yum_repository epel; then
    yum -y install epel-release
    # disable it by default!
    epel_repo=/etc/yum.repos.d/epel.repo
    backup=/tmp/$(basename $epel_repo).bak
    mv $epel_repo $backup
    cat $backup | $sed -e "s/enabled=1/enabled=0/" \
      > $epel_repo
  fi

  yum -y --enablerepo=epel makecache
  yum -y install curl
  yum -y --enablerepo=epel install npm

  if [ "$VERSION" != "release" ]; then
    yum -y install git
  fi
}
# ========================= /for CentOS 7 ===========================



install() {
  mkdir -p $TEMPDIR

  echo "Preparing the environment..."
  prepare_environment_in_$PLATFORM

  echo ""
  if [ "$VERSION" != "release" ]; then
    echo "Installing $NAME from the git repository..."
    install_from_repository
  else
    echo "Installing $NAME from npmjs.org..."
    npm install -g droonga-http-server
  fi

  if ! exist_command droonga-http-server; then
    echo "ERROR: Failed to install $NAME!"
    exit 1
  fi

  curl -s -o $TEMPDIR/functions.sh $(download_url "install/$PLATFORM/functions.sh")
  if ! source $TEMPDIR/functions.sh; then
    echo "ERROR: Failed to download post-installation script!"
    exit 1
  fi
  if ! exist_command register_service; then
    echo "ERROR: Downloaded post-installation script is broken!"
    exit 1
  fi

  prepare_user

  setup_configuration_directory

  echo ""
  echo "Registering $NAME as a service..."
  # this function is defined by the downloaded "functions.sh"!
  register_service $NAME $USER $GROUP

  echo ""
  echo "Successfully installed $NAME."
}

ensure_root

PLATFORM=$(guess_platform)
if [ "$PLATFORM" = "" ]; then
  echo "Not supported platform."
  exit 255
fi

install

exit 0


application.js
var express = require('express'),
    droonga = require('express-droonga'),
    responseTime = require('response-time'),
    http = require('http'),
    options = require('commander');

var version = require('./package.json').version;

options
  .version(version)
  .option('--port <port>', 'Port number', parseInt, 13000)
  .option('--receive-host-name <name>',
          'Host name of the protocol adapter. ' +
            'It must be resolvable by Droonga engine.',
          '127.0.0.1')
  .option('--droonga-engine-host-name <name>', 'Host name of Droonga engine',
          '127.0.0.1')
  .option('--droonga-engine-port <port>', 'Port number of Droonga engine',
          parseInt, 24224)
  .option('--default-dataset <dataset>', 'The default dataset',
          'Droonga')
  .option('--enable-logging', 'Enable logging to the standard output')
  .option('--cache-size <size>', 'The max number of cached requests',
          parseInt, 100)
  .parse(process.argv);

var application = express();
var server = http.createServer(application);

var MemoryStore = express.session.MemoryStore;
var sessionStore = new MemoryStore();
application.configure(function() {
  if (options.enableLogging) {
    application.use(express.logger());
  }
  application.use(express.cookieParser('secret key'));
  application.use(express.session({
    secret: 'secret key',
    store:  sessionStore
  }));
  application.use(responseTime());
  if (options.cacheSize > 0) {
    application.use(droonga.cache({
      size: options.cacheSize,
      rules: [
        { regex: /./ }
      ]
    }));
  }
});

application.droonga({
  prefix: '',
  defaultDataset: options.defaultDataset,
  server: server,
  sessionStore: sessionStore, // this is required to share session information HTTP APIs
  receiveHostName: options.receiveHostName,
  hostName: options.droongaEngineHostName,
  port: options.droongaEnginePort,
  plugins: [
    droonga.API_REST,
    droonga.API_GROONGA,
    droonga.API_DROONGA
  ]
});

server.listen(options.port);


lib/serf/client.js
/**
 * usage:
 *   var Client = require('lib/serf/client');
 *   var serf = new Client({
 *                host:       'node0', // host name that the serf agent is working
 *                tag:        'droonga', // tag of engines, can be omitted
 *                enginePort: 10031 // port number of engines, can be omitted
 *              });
 *   serf.getAllMembers()
 *         .then(function(members) { console.log(members); })
 *         // members:
 *         //   [{ Name:     'node0:10031/droonga',
 *         //      Port:     7946,
 *         //      Status:   'alive', ... }, ...]
 *         .catch(function(e) { console.error(e); });
 *   serf.getLiveEngineNodes()
 *         .then(function(members) { console.log(members); })
 *         // members:
 *         //   [{ Name:     'node0:10031/droonga',
 *         //      Port:     7946,
 *         //      Status:   'alive',
 *         //      HostName: 'node0', ... }, ...]
 *         .catch(function(e) { console.error(e); });
 */

var SerfRPC = require('serf-rpc'),
    Q       = require('q');

var ENGINE_NODE_NAME_PATTERN = /^([^:]+):(\d+)\/(.+)$/;

function Client(options) {
  options = options || {};

  this._droongaHost       = options.host || '127.0.0.1';
  this._droongaEnginePort = options.enginePort || 10031;
  this._droongaTag        = options.tag || 'droonga';

  this._connectionOptions = {
    rpc_host: this._droongaHost,
    rpc_port: 7373
  };
  this._client = new SerfRPC();
  this._connected = false;
}
Client.prototype = {
  connect: function() {
    return Q.Promise((function(resolve, reject, notify) {
      if (this._connected)
        return resolve();

      this._client.connect(this._connectionOptions, (function(error) {
        if (error)
          return reject(error);
        this._connected = true;
        resolve();
      }).bind(this));
    }).bind(this));
  },

  joinTo: function(existingMemberHostName) {
    return this.connect().then((function() {
      return Q.Promise((function(resolve, reject, notify) {
        this._client.join({
          Existing: [existingMemberHostName + ':7946'],
          Replay:   false
        }, function(error, result) {
          if (error)
            return reject(error);
          resolve();
        });
      }).bind(this));
    }).bind(this));
  },

  getAllMembers: function() {
    return this.connect().then((function() {
      return Q.Promise((function(resolve, reject, notify) {
        this._client.members(function(error, result) {
          if (error)
            return reject(error);
          resolve(result.Members);
        });
      }).bind(this));
    }).bind(this));
  },

  getLiveEngineNodes: function() {
    return this.getAllMembers().then((function(members) {
      return members.filter(function(member) {
        var matched = String(member.Name).match(ENGINE_NODE_NAME_PATTERN);
        if (matched)
          member.HostName = matched[1];
        return matched &&
                 matched[2] == this._droongaPort &&
                 matched[3] == this._droongaTag &&
                 member.Status == 'alive';
      }, this);
    }).bind(this));
  }
};

module.exports = Client;


lib/option-utils.js
var LIST_SEPARATOR_MATCHER = /\s*[,|]\s*/;

function normalizeStringArray(values) {
  if (Array.isArray(values))
    return values;

  values = values.trim();
  if (!values)
    return [];

  if (LIST_SEPARATOR_MATCHER.test(values))
    return values.split(LIST_SEPARATOR_MATCHER);
  else
    return values.split(/\s+/);
}
exports.normalizeStringArray = normalizeStringArray;


function intOption(newValue, oldValue) {
  return parseInt(newValue);
}
exports.intOption = intOption;

function floatOption(newValue, oldValue) {
  return parseFloat(newValue);
}
exports.floatOption = floatOption;

function stringsOption(newValue, oldValue) {
  return normalizeStringArray(newValue);
}
exports.stringsOption = stringsOption;

function pluginsOption(newValue, oldValue) {
  return normalizeStringArray(newValue).map(function (plugin) {
    return require(plugin);
  });
}
exports.pluginsOption = pluginsOption;


lib/server-options.js
var options = require('commander');

var version = require('../package.json').version;
var defaultConfigs = require('./default-configs'),
    optionUtils    = require('./option-utils');

options.host             = defaultConfigs.host;
options.port             = defaultConfigs.port;
options.accessLogFile    = defaultConfigs.access_log_file;
options.systemLogFile    = defaultConfigs.system_log_file;
options.systemLogLevel   = defaultConfigs.system_log_level;
options.daemon           = defaultConfigs.daemon;
options.pidFile          = defaultConfigs.pid_file
options.cacheSize        = defaultConfigs.cache_size;
options.cacheTtlInSeconds = defaultConfigs.cache_ttl_in_seconds;
options.enableTrustProxy = defaultConfigs.enable_trust_proxy;
options.documentRoot     = defaultConfigs.document_root;
options.plugins          = defaultConfigs.plugins;
options.environment      = defaultConfigs.environment;

options.droongaEngineHostNames = defaultConfigs.engine.hosts;
options.droongaEnginePort      = defaultConfigs.engine.port;
options.tag                    = defaultConfigs.engine.tag;
options.defaultDataset         = defaultConfigs.engine.default_dataset;
options.receiveHostName        = defaultConfigs.engine.receive_host;
options.defaultTimeout         = defaultConfigs.engine.default_timeout;

function generateOptionHandler(onHandle, converter) {
  return function(newValue, oldValue) {
    onHandle(newValue);
    if (converter)
      return converter(newValue);
    else
      return newValue;
  };
}


options = options
  .version(version);

function add() {
  options = options.option.apply(options, arguments);
  return exports;
}
exports.add = add;

function define() {
  add('--host <host>',
      'Host to listen (' + options.host + ')');
  add('--port <port>',
      'Port number (' + options.port + ')',
      optionUtils.intOption);
  add('--receive-host-name <name>',
      'Host name of the protocol adapter. ' +
        'It must be resolvable by Droonga engine. ' +
        '(' + options.receiveHostName + ')');
  add('--droonga-engine-host-names <name1,name2,...>',
      'List of Droonga engine nodes\' host name (' + options.droongaEngineHostNames.join(',') + ')',
      optionUtils.stringsOption);
  add('--droonga-engine-host-name <name1,name2,...>',
      'Alias of --droonga-engine-host-names for backward compatibility.',
      optionUtils.stringsOption);
  add('--droonga-engine-port <port>',
      'Port number of Droonga engine (' + options.droongaEnginePort + ')',
      optionUtils.intOption);
  add('--default-dataset <dataset>',
      'The default dataset (' + options.defaultDataset + ')');
  add('--tag <tag>',
      'The tag (' + options.tag + ')');
  add('--default-timeout <timeout>',
      'Time to terminate unresponsive connections (in seconds) (' + options.defaultTimeout + ')',
      optionUtils.floatOption);
  add('--access-log-file <file>',
      'Output access logs to <file>. ' +
        'You can use "-" as <file> to output to the standard output. ' +
        '(' + options.accessLogFile + ')');
  add('--system-log-file <file>',
      'Output system logs to <file>. ' +
        'You can use "-" as <file> to output to the standard output. ' +
        '(' + options.systemLogFile + ')');
  add('--system-log-level <level>',
      'Log level for the system log. ' +
      '[silly,debug,verbose,info,warn,error]' +
        '(' + options.systemLogLevel + ')');
  add('--cache-size <size>',
      'The maximum number of cached responses ' +
        '(' + options.cacheSize + ')',
      optionUtils.intOption);
  add('--cache-ttl-in-seconds <seconds>',
      'The time to live of cached responses, in seconds ' +
        '(' + options.cacheTtlInSeconds + ')',
      optionUtils.intOption);
  add('--enable-trust-proxy',
      'Enable "trust proxy" configuration. It is required when you run droonga-http-server behind a reverse proxy. ' +
        '(' + options.enableTrustProxy + ')');
  add('--disable-trust-proxy',
      'Inverted option of --enable-trust-proxy.');
  add('--document-root <path>',
      'Path to the document root. ' +
        '(' + options.documentRoot + ')');
  add('--plugins <plugin1,plugin2,...>',
      'Use specified plugins. ' +
        '(' + options.plugins.join(',') + ')',
      optionUtils.pluginsOption);
  add('--daemon',
      'Run as a daemon. (' + options.daemon + ')');
  add('--pid-file <pid-file>',
      'Output PID to <pid-file>.');
  add('--environment <environment>',
      'Use specified environment. (' + options.environment + ')');
  return exports;
}
exports.define = define;

function parse(argv) {
  var parsedOptions = options.parse(argv);

  if (parsedOptions.disableTrustProxy)
    parsedOptions.enableTrustProxy = false;

  // for backward compatibility
  if (options.droongaEngineHostName &&
      options.droongaEngineHostName.length > 0 &&
      (!options.droongaEngineHostNames ||
       options.droongaEngineHostNames.length == 0))
    options.droongaEngineHostNames = options.droongaEngineHostName;

  return parsedOptions;
}
exports.parse = parse;


install/centos/functions.sh
# Copyright (C) 2014 Droonga Project
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License version 2.1 as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

register_service() {
  NAME=$1
  USER=$2
  GROUP=$3

  #TODO: we should migrate to systemd in near future...

  pid_dir=/run/$NAME
  mkdir -p $pid_dir
  chown -R $USER:$GROUP $pid_dir

  curl -o /etc/rc.d/init.d/$NAME $(download_url "install/centos/$NAME")
  chmod +x /etc/rc.d/init.d/$NAME
  /sbin/chkconfig --add $NAME
}


lib/default-configs.js
var droonga = require('express-droonga'),
    fs      = require('fs'),
    yaml    = require('js-yaml'),
    path    = require('path');

var optionUtils = require('./option-utils');

var baseDir = path.resolve(process.env.DROONGA_BASE_DIR || '.');

var engineConfigs = {};
var engineServiceUserName = 'droonga-engine';
var engineServiceBaseDir  = '/home/' + engineServiceUserName + '/droonga';

function setBaseDir(baseDir) {
  var configs = {};

  var configFile = path.resolve(baseDir, 'droonga-http-server.yaml');
  if (fs.existsSync(configFile)) {
    configs = yaml.safeLoad(fs.readFileSync(configFile, 'utf8'));
  }

  var engineServiceConfigFile = path.resolve(engineServiceBaseDir, 'droonga-engine.yaml');
  if (fs.existsSync(engineServiceConfigFile)) {
    engineConfigs = yaml.safeLoad(fs.readFileSync(engineServiceConfigFile, 'utf8'));
  }
  else {
    var engineConfigFile = path.resolve(baseDir, 'droonga-engine.yaml');
    if (fs.existsSync(engineConfigFile)) {
      engineConfigs = yaml.safeLoad(fs.readFileSync(engineConfigFile, 'utf8'));
    }
  }

  configs.baseDir = baseDir;
  configs = defineDefaultConfigs(configs);
  configs.setBaseDir = setBaseDir;
  return configs;
}

function define(slot, path, value) {
  var keys = path.split('.');
  keys.some(function(key, index) {
    if (index == keys.length - 1) {
      if (!(key in slot))
        slot[key] = value;
    } else {
      if (!(key in slot))
        slot[key] = {};
      slot = slot[key];
    }
  });
}

function defineDefaultConfigs(configs) {
  define(engineConfigs, 'host', '127.0.0.1');
  define(engineConfigs, 'port', 10031);
  define(engineConfigs, 'tag',  'droonga');
  define(engineConfigs, 'default_timeout', 3);

  define(configs, 'host',               '0.0.0.0');
  define(configs, 'port',               10041);
  define(configs, 'cache_size',         100);
  define(configs, 'cache_ttl_in_seconds', 60);
  define(configs, 'enable_trust_proxy', false);
  define(configs, 'plugins', [
    droonga.API_REST,
    droonga.API_GROONGA,
    droonga.API_DROONGA
  ]);
  define(configs, 'document_root', path.normalize(__dirname + '/../public/groonga-admin'));
  define(configs, 'environment', process.env.NODE_ENV || 'development');

  // Ignore value of "daemon" option defined in the configuration file,
  // because "commander" is not designed for a boolean option with
  // unfixed default value.
  configs.daemon = false;
  // define(configs, 'daemon',          false);
  define(configs, 'access_log_file', '-');
  define(configs, 'system_log_file', '-');
  define(configs, 'system_log_level', 'warn');

  // for backward compatibility
  define(configs, 'engine.host',            engineConfigs.host);

  define(configs, 'engine.hosts',           optionUtils.normalizeStringArray(configs.engine.host));
  configs.engine.host = optionUtils.normalizeStringArray(configs.engine.hosts);

  define(configs, 'engine.port',            engineConfigs.port);
  define(configs, 'engine.tag',             engineConfigs.tag);
  define(configs, 'engine.default_dataset', 'Default');
  // We can use the host name of the droonga-engine as the receive host
  // of the http server, because this computer works as a droonga-engine
  // node and it is guaranteed that this computer can be accessed with
  // the host name.
  define(configs, 'engine.receive_host',    engineConfigs.host);
  define(configs, 'engine.default_timeout', engineConfigs.default_timeout);

  return configs;
}

module.exports = setBaseDir(baseDir);
